### PR TITLE
[bugfix] TPU test hangs to barrier on 1 process

### LINKED ISSUE #6230: TPU:  Crashes using trainer.test()
# 🐛 Bug

`trainer.test()` does not work with TPUs.

There are a few different ways we've seen it crash.

### 1. Looks like a call to `barrier()` coming from `__test_using_best_weights`
```
RuntimeError                              Traceback (most recent call last)
<ipython-input-17-587e2a9e3858> in <module>
----> 1 trainer.test(datamodule=dm)

/opt/conda/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py in test(self, model, test_dataloaders, ckpt_path, verbose, datamodule)
    922             results = self.__test_given_model(model, test_dataloaders)
    923         else:
--> 924             results = self.__test_using_best_weights(ckpt_path, test_dataloaders)
    925 
    926         self.teardown('test')

/opt/conda/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py in __test_using_best_weights(self, ckpt_path, test_dataloaders)
    950                 return {}
    951             if not self._device_type == DeviceType.TPU:
--> 952                 self.accelerator.barrier()
    953 
    954             ckpt = pl_load(ckpt_path, map_location=lambda storage, loc: storage)

/opt/conda/lib/python3.7/site-packages/pytorch_lightning/accelerators/accelerator.py in barrier(self, name)
    375 
    376     def barrier(self, name: Optional[str] = None) -> None:
--> 377         self.training_type_plugin.barrier(name=name)
    378 
    379     def broadcast(self, obj: object, src: int = 0) -> object:

/opt/conda/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/tpu_spawn.py in barrier(self, name)
    113 
    114     def barrier(self, name: Optional[str] = None) -> None:
--> 115         rendezvous(f"pl.Trainer.{name}")
    116 
    117     def transfer_distrib_spawn_state_on_fit_end(self, results):

/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py in rendezvous(tag, payload, replicas)
    859     ordinal `i` at position `i` in the returned tuple.
    860   """
--> 861   return torch_xla._XLAC._xla_rendezvous(get_ordinal(), tag, payload, replicas)
    862 
    863 

RuntimeError: tensorflow/compiler/xla/xla_client/mesh_service.cc:316 : Check failed: impl_->channel->WaitForConnected( std::chrono::system_clock::now() + std::chrono::seconds(connect_wait_seconds))  
```
So the barrier is coming from [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#L915). This is strange that `barrier` is being called - I think this means that `if not self._device_type == DeviceType.TPU` is mistakenly evaluating to `True`?  I think pytorch lightning spins up 8 processes for 8 TPU cores, is it possible only some of them are evaluating to `True`?

Basically it seems like at least 1 process is not making it to this point, which means the other processes are waiting in the barrier and the meetup never happens so we get the `RuntimeError` shown.

### 2. Looks like a call to `xm.save()` is being misused:
```
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 330, in _mp_start_fn
    _start_fn(index, pf_cfg, fn, args)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 330, in _mp_start_fn
    _start_fn(index, pf_cfg, fn, args)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 324, in _start_fn
    fn(gindex, *args)
  File "/opt/conda/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/tpu_spawn.py", line 103, in new_process
    self.transfer_distrib_spawn_state_on_fit_end(results)
  File "/opt/conda/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/tpu_spawn.py", line 129, in transfer_distrib_spawn_state_on_fit_end
    xm.save(self.lightning_module.state_dict(), last_path)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py", line 817, in save
    rendezvous('torch_xla.core.xla_model.save')
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py", line 861, in rendezvous
    return torch_xla._XLAC._xla_rendezvous(get_ordinal(), tag, payload, replicas)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/distributed/xla_multiprocessing.py", line 324, in _start_fn
    fn(gindex, *args)
RuntimeError: tensorflow/compiler/xla/xla_client/mesh_service.cc:364 : Failed to meet rendezvous 'torch_xla.core.xla_model.save': Socket closed (14)
  File "/opt/conda/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/tpu_spawn.py", line 103, in new_process
    self.transfer_distrib_spawn_state_on_fit_end(results)
  File "/opt/conda/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/tpu_spawn.py", line 129, in transfer_distrib_spawn_state_on_fit_end
    xm.save(self.lightning_module.state_dict(), last_path)
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py", line 817, in save
    rendezvous('torch_xla.core.xla_model.save')
  File "/opt/conda/lib/python3.7/site-packages/torch_xla/core/xla_model.py", line 861, in rendezvous
    return torch_xla._XLAC._xla_rendezvous(get_ordinal(), tag, payload, replicas)
Exception in device=TPU:6: tensorflow/compiler/xla/xla_client/mesh_service.cc:364 : Failed to meet rendezvous 'torch_xla.core.xla_model.save': Socket closed (14)Exception in device=TPU:3: tensorflow/compiler/xla/xla_client/mesh_service.cc:364 : Failed to meet rendezvous 'torch_xla.core.xla_model.save': Socket closed (14)
```
I think the problem is [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/tpu_spawn.py#L140) with the usage of `xm.save()`.

`xm.save()` already handles the multiprocess case by checking the ordinal and only writing to disk if the process is on the master ordinal.  In general, if you surround `xm.save()` with `if` statements, it means some TPU cores enter the `if` statement and some will not, so the cores that entered the `if` statement will be waiting for those that didn't enter and eventually it will time out and crash.

# Repro methods

### 1. (Colab) Make 3 modifications to the  [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing)
1. Switch runtime version to TPU
2. Add this cell as the first cell:
```
VERSION = "1.7"  #@param ["1.7" , "20200516", "nightly"]
!curl https://raw.githubusercontent.com/pytorch/xla/master/contrib/scripts/env-setup.py -o pytorch-xla-env-setup.py
!python pytorch-xla-env-setup.py --version $VERSION
```
3. Add `tpu_cores=8` to the trainer cell

### 2. (Google Cloud) Use the attached repro.py file in the following way:
1. create a TPU
2. create a Google Cloud VM (I used e2-standard-32 but size shouldn't matter too much)
3. SSH into VM
4. (VM) `conda activate torch-xla-1.7`
5. (VM) `pip install pytorch-lightning==1.2.1`
5. (VM) `export TPU_IP_ADDRESS=my.tpu.ip.addr`
6. (VM) `export XRT_TPU_CONFIG="tpu_worker;0;$TPU_IP_ADDRESS:8470"`
7. (VM) `python3 repro.py`

### 3. (Your CI setup) Modify TPU unit tests as follows:
* Add a `trainer.test(test_dataloaders=DataLoader(RandomDataset(32, 2000), batch_size=32))` after some call to `trainer.fit`
* For example, I changed test_model_tpu_early_stop test to look like this:
```
@pytest.mark.skipif(not _TPU_AVAILABLE, reason="test requires TPU machine")
@pl_multi_process_test
def test_model_tpu_early_stop(tmpdir):
    """Test if single TPU core training works"""

    # todo: Test on 8 cores - hanging.

    class CustomBoringModel(BoringModel):

        def validation_step(self, *args, **kwargs):
            out = super().validation_step(*args, **kwargs)
            self.log('val_loss', out['x'])
            return out

    tutils.reset_seed()
    model = CustomBoringModel()
    trainer = Trainer(
        callbacks=[EarlyStopping(monitor='val_loss')],
        default_root_dir=tmpdir,
        progress_bar_refresh_rate=0,
        max_epochs=2,
        limit_train_batches=2,
        limit_val_batches=2,
        tpu_cores=[1],
    )
    trainer.fit(model)
+    trainer.test(test_dataloaders=DataLoader(RandomDataset(32, 2000), batch_size=32))
```

Ran tests with `coverage run --source=pytorch_lightning -m pytest tests/models/test_tpu.py -v`. This should allow testing on the CI framework

### Environment

 - PyTorch Version (e.g., 1.0): 1.7
 - OS (e.g., Linux): Linux
 - Build command you used (if compiling from source): `pip install pytorch-lightning==1.2.1` (note that earlier versions hang due to #5841 )
 - Python version: 3.6
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR resolves a bug with TPU.

Fixes #6230

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -911,8 +911,9 @@ def __test_using_best_weights(self, ckpt_path, test_dataloaders):
                     f'specify a path for a checkpoint .test(ckpt_path=PATH)'
                 )
                 return {}
-            if not self._device_type == DeviceType.TPU:
-                self.accelerator.barrier()
+
+            if torch.distributed.is_available() and torch.distributed.is_initialized():
+                self.training_type_plugin.barrier()
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:916]: would it make sense to move the check for `torch.distributed` into the specific barrier method in the plugin?
Then only the plugins that use distributed would run this check and others will do nothing

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 2]
```diff
@@ -139,14 +141,26 @@ def transfer_distrib_spawn_state_on_fit_end(self, results):
...
+        """
+        Saving with xm.save can failed to meet rendez-vous.
+        Therefore, we will try several times to do so.
+        """
+        for _ in range(self._repeat_save_on_fail):
+            try:
+                xm.save(state_dict, path)
+                break
+            except RuntimeError:
+                sleep(0.001)
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:162]: how about the last try, there is no final error if all fails?

File: tests/models/test_tpu.py [hunk 3]
```diff
@@ -195,9 +193,10 @@ def validation_step(self, *args, **kwargs):
         max_epochs=2,
         limit_train_batches=2,
         limit_val_batches=2,
-        tpu_cores=[1],
+        tpu_cores=8,
     )
     trainer.fit(model)
+    trainer.test(test_dataloaders=DataLoader(RandomDataset(32, 2000), batch_size=32))
```
Comment (zcain117) [inline tests/models/test_tpu.py:199]: Thanks for updating this test too

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:916]: Good idea !

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py]: I moved it into a try / catch instead.
I still don't know why it is so unstable, but it shouldn't affect training.

**Review by awaelchli — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 4]
```diff
@@ -139,14 +141,28 @@ def transfer_distrib_spawn_state_on_fit_end(self, results):
...
+        The rendez-vous doesn't affect directly saving.
+        We can ignore the ``RuntimeError`` to reduce friction with TPUs.
+        """
+        try:
+            xm.save(state_dict, path)
+        except RuntimeError as e:
+            if "Failed to meet rendezvous" in str(e):
+                pass
+            else:
+                raise e
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:164]:
```suggestion
            if "Failed to meet rendezvous" not in str(e):
                raise e
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 5]
```diff
@@ -139,14 +141,26 @@ def transfer_distrib_spawn_state_on_fit_end(self, results):
...
-                xm.save(self.lightning_module.state_dict(), last_path)
+                self.save(self.lightning_module.state_dict(), last_path)
 
             if self.global_rank == 0:
                 # todo, pass complete checkpoint as state dictionary
                 self.mp_queue.put(best_model_path)
                 self.mp_queue.put(last_path)
                 self.mp_queue.put(results)
 
+    def save(self, state_dict: Dict, path: str) -> None:
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:152]: shall this be rather protected?
```suggestion
    def _save(self, state_dict: Dict, path: str) -> None:
```

**Review by Borda — APPROVED** [review]

**Review by williamFalcon — APPROVED** [review]

**Pull request merged**